File: ios-native/Sources/ALUI/Engines/TwinsModel.swift
import ALCore
import Observation

// The multi-select twins engine — `SoundTwinsExercise.tsx` (engines.md §4.3).
//
// Hear one sound (« Trouve tous les ko ! »), tap EVERY tile that writes it.
// A correct partial tap locks its tile into the collection strip and speaks
// its anchor word FIRE-AND-FORGET (the next tap can land while the line
// plays); the COMPLETING tap locks the round and awaits its line before
// advancing. An intruder tap follows the single-pick miss rules: nudge +
// cooldown + the round's star greys at pointerdown. Partial finds never grey
// the star and never change the strip — only intruders do.

@MainActor
@Observable
public final class TwinsModel {
    @ObservationIgnored public let level: Int
    /// Seeded ONCE, in this init (D9 — never build the model in a view `init`
    /// or a `@State` default).
    public let session: [TwinRound]

    public private(set) var idx = 0
    /// This round's found tile ids, in tap order — fills the collection strip.
    public private(set) var found: [Int] = []
    public private(set) var mood: Mood = .idle
    public private(set) var done = false
    public private(set) var earned = 0
    public private(set) var stars: [Bool]

    @ObservationIgnored private var locked = false
    @ObservationIgnored private var isActive = true
    @ObservationIgnored private let cooldown: MissCooldown
    @ObservationIgnored private let deps: EngineDeps
    @ObservationIgnored private var announceTask: Task<Void, Never>?

    public init(
        level: Int,
        deps: EngineDeps,
        rng: RandomSource = .system()
    ) {
        self.level = level
        self.deps = deps
        self.cooldown = MissCooldown(time: deps.time)
        let session = Levels.buildTwinSession(level: level, rng)
        self.session = session
        self.stars = session.map { _ in true }
        self.done = session.isEmpty
    }

    // MARK: - Projection surface

    public var round: TwinRound { session[idx] }

    /// The twins to find — one collection-strip slot each.
    public var targets: [TwinTile] { round.tiles.filter(\.correct) }

    /// Every target found — the round is celebrating (or about to advance).
    public var complete: Bool {
        let found = self.found
        return targets.allSatisfy { found.contains($0.id) }
    }

    /// A found tile locks; a COMPLETE round locks the rest too, so nothing
    /// shakes while the celebration line plays out.
    public func tileDisabled(_ tile: TwinTile) -> Bool {
        found.contains(tile.id) || complete
    }

    /// Found tiles wear the green ring.
    public func tileHighlighted(_ tile: TwinTile) -> Bool {
        found.contains(tile.id)
    }

    /// Collection strip slot `i`: the i-th found tile, or nil (dashed box).
    public func foundTile(at i: Int) -> TwinTile? {
        guard found.indices.contains(i) else { return nil }
        let id = found[i]
        return round.tiles.first { $0.id == id }
    }

    // MARK: - Lifecycle

    public func activate() {
        isActive = true
        deps.audio.unlock()
        scheduleAnnounce()
    }

    public func deactivate() {
        announceTask?.cancel()
        announceTask = nil
        isActive = false
        deps.audio.stop()
    }

    // MARK: - The pick handler (synchronous — invariant 1)

    public func pick(_ tile: TwinTile) -> Verdict {
        if done { return .reject }  // unreachable from a correct view
        if locked { return .reject }
        if cooldown.isSwallowing { return .reject }  // silent while the shake plays
        deps.audio.unlock()
        deps.audio.pop()
        if !tile.correct {
            deps.audio.nudge()
            cooldown.registerMiss()
            StarStrip.miss(idx, in: &stars)  // the star greys NOW, same beat as the shake
            return .reject
        }
        found.append(tile.id)
        mood = .happy
        // The anchor word IS the feedback, per graphy.
        let line = Levels.twinSuccess(
            TwinGraphy(text: tile.text, word: tile.word, emoji: tile.emoji))
        if !complete {
            // A twin found, more to go — fire-and-forget; the next tap can
            // land while this line plays (say is latest-wins).
            let audio = deps.audio
            Task {
                _ = await audio.say(
                    line, rate: EngineLines.successRate, pitch: EngineLines.defaultPitch)
            }
            return .accept
        }
        locked = true
        deps.audio.success()
        deps.fireConfetti()
        // Advance only after the LAST success line played in full (`ok`).
        let next = idx + 1
        Task { [weak self] in
            guard let self else { return }
            let ok = await self.deps.audio.say(
                line, rate: EngineLines.successRate, pitch: EngineLines.defaultPitch)
            guard ok, self.isActive else { return }
            self.locked = false
            if next >= self.session.count {
                self.announceTask?.cancel()
                self.mood = .cheer
                self.earned = self.deps.award(
                    .soundTwins, self.level, self.stars.filter { $0 }.count, self.session.count)
                self.done = true
                let audio = self.deps.audio
                Task { _ = await audio.say(EngineLines.bravoFound) }
            } else {
                self.mood = .idle
                self.found = []
                self.idx = next
                self.scheduleAnnounce()
            }
        }
        return .accept
    }

    // MARK: - Audio affordances

    /// The big 🔊 button — locked-guarded, no unlock (TSX shape).
    public func replayPrompt() {
        guard !locked, !done, !session.isEmpty else { return }
        let line = Levels.twinPrompt(session[idx].family)
        let audio = deps.audio
        Task { _ = await audio.say(line) }
    }

    /// A tile's « Écouter » — speaks the tile's own family's sound
    /// (`tile.sound`); locked-guarded in the TSX.
    public func preview(say text: String) {
        guard !locked else { return }
        deps.audio.unlock()
        let audio = deps.audio
        Task { _ = await audio.say(text) }
    }

    // MARK: - Announce (350 ms after every round becomes current)

    private func scheduleAnnounce() {
        announceTask?.cancel()
        guard !done, !session.isEmpty else { return }
        let expected = idx
        let line = Levels.twinPrompt(session[expected].family)
        announceTask = Task { [weak self] in
            guard let self else { return }
            await self.deps.delay(EngineLines.announceDelayMs)
            guard !Task.isCancelled, self.isActive, !self.done, self.idx == expected
            else { return }
            _ = await self.deps.audio.say(line)
        }
    }
}

extension TwinsModel: RoundRunner {
    public var totalRounds: Int { session.count }
    /// U+0027 apostrophe + U+2014 em dash — never normalised (D17).
    public var headline: String? {
        "Un son peut s'écrire de plusieurs façons — trouve-les toutes !"
    }
    public var finishedTitle: String { Copy.Finished.allFound }
    public var listenAccessibilityLabel: String { Copy.Exercise.replaySound }
}

File: apps/game-android/ui/src/main/kotlin/fr/dappit/attrapelettres/ui/engines/TwinsModel.kt
package fr.dappit.attrapelettres.ui.engines

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import fr.dappit.attrapelettres.core.domain.ExerciseId
import fr.dappit.attrapelettres.core.domain.Mood
import fr.dappit.attrapelettres.core.domain.TwinGraphy
import fr.dappit.attrapelettres.core.domain.TwinRound
import fr.dappit.attrapelettres.core.domain.TwinTile
import fr.dappit.attrapelettres.core.domain.Verdict
import fr.dappit.attrapelettres.core.levels.buildTwinSession
import fr.dappit.attrapelettres.core.levels.twinPrompt
import fr.dappit.attrapelettres.core.levels.twinSuccess
import fr.dappit.attrapelettres.core.support.RandomSource
import fr.dappit.attrapelettres.core.support.SystemRandomSource
import fr.dappit.attrapelettres.ui.components.MissCooldown
import fr.dappit.attrapelettres.ui.components.StarStrip
import fr.dappit.attrapelettres.ui.design.Copy
import kotlinx.coroutines.Job
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

// The multi-select twins engine — `SoundTwinsExercise.tsx`.
//
// Hear one sound (« Trouve tous les ko ! ») and tap EVERY tile that writes it.
// A correct PARTIAL tap locks its tile into the collection strip and speaks its
// anchor word FIRE-AND-FORGET — the next tap can land while that line plays,
// because a child who has spotted two twins should not have to wait to say so.
// The COMPLETING tap locks the round and awaits its line before advancing.
//
// An intruder tap follows the single-pick miss rules exactly: nudge + cooldown
// + the round's star greys at pointer-down. Partial finds never grey the star
// and never touch the strip — only intruders do (invariants 3 and 8).

/**
 * The loop. See [SinglePickModel] for why `@Stable` is honest here.
 *
 * The exercise id is fixed rather than injected: `SOUND_TWINS` is the only
 * catalog row this engine serves, and hard-coding it keeps the award call —
 * the one place a wrong id would silently move a child's stars to another
 * ledger key — impossible to get wrong from a screen.
 */
@Stable
class TwinsModel(
    val level: Int,
    private val deps: EngineDeps,
    rng: RandomSource = SystemRandomSource(),
) : RoundRunner {

    /**
     * Seeded ONCE, here — build the model inside a `remember`, never in a
     * composable body (see RoundRunner.kt's header).
     */
    val session: List<TwinRound> = buildTwinSession(level, rng)

    private var idxState by mutableStateOf(0)
    private var foundState: List<Int> by mutableStateOf(emptyList())
    private var moodState by mutableStateOf(Mood.IDLE)
    private var doneState by mutableStateOf(session.isEmpty())
    private var earnedState by mutableStateOf(0)
    private val starList = mutableStateListOf<Boolean>().apply {
        repeat(session.size) { add(true) }
    }

    private var locked = false
    private var mounted = true
    private val cooldown = MissCooldown(deps.time)
    private var announceJob: Job? = null

    // --- Projection surface --------------------------------------------------

    override val totalRounds: Int get() = session.size
    override val idx: Int get() = idxState
    override val stars: List<Boolean> get() = starList
    override val mood: Mood get() = moodState
    override val done: Boolean get() = doneState
    override val earned: Int get() = earnedState

    /** U+0027 apostrophe + U+2014 em dash — never normalised. */
    override val headline: String get() = Copy.Exercise.SOUND_TWINS_HEADLINE
    override val finishedTitle: String get() = Copy.Finished.ALL_FOUND
    override val listenAccessibilityLabel: String get() = Copy.Exercise.REPLAY_SOUND

    val round: TwinRound get() = session[idxState]

    /** This round's found tile ids, in tap order — fills the collection strip. */
    val found: List<Int> get() = foundState

    /** The twins to find — one collection-strip slot each. */
    val targets: List<TwinTile> get() = round.tiles.filter { it.correct }

    /** Every target found: the round is celebrating (or about to advance). */
    val complete: Boolean
        get() {
            val ids = foundState
            return targets.all { ids.contains(it.id) }
        }

    /**
     * A found tile locks; a COMPLETE round locks the rest too, so nothing
     * shakes while the celebration line plays out.
     */
    fun tileDisabled(tile: TwinTile): Boolean = foundState.contains(tile.id) || complete

    /** Found tiles wear the green ring. */
    fun tileHighlighted(tile: TwinTile): Boolean = foundState.contains(tile.id)

    /** Collection strip slot `i`: the i-th found tile, or null (a dashed box). */
    fun foundTile(i: Int): TwinTile? {
        val ids = foundState
        if (i !in ids.indices) return null
        return round.tiles.firstOrNull { it.id == ids[i] }
    }

    // --- Lifecycle -----------------------------------------------------------

    override fun activate() {
        mounted = true
        deps.audio.unlock()
        scheduleAnnounce()
    }

    override fun deactivate() {
        announceJob?.cancel()
        announceJob = null
        mounted = false
        deps.audio.stop()
    }

    // --- The pick handler (synchronous — invariant 1) ------------------------

    fun pick(tile: TwinTile): Verdict {
        if (doneState) return Verdict.REJECT // unreachable from a correct screen
        if (locked) return Verdict.REJECT
        if (cooldown.isSwallowing) return Verdict.REJECT // silent while the shake plays
        deps.audio.unlock()
        deps.audio.pop()
        if (!tile.correct) {
            deps.audio.nudge()
            cooldown.registerMiss()
            StarStrip.miss(idxState, starList) // the star greys NOW, same beat as the shake
            return Verdict.REJECT
        }
        foundState = foundState + tile.id
        moodState = Mood.HAPPY
        // The anchor word IS the feedback, per graphy.
        val line = twinSuccess(TwinGraphy(text = tile.text, word = tile.word, emoji = tile.emoji))
        if (!complete) {
            // A twin found, more to go — fire-and-forget; the next tap can land
            // while this line plays (`say` is latest-wins).
            deps.scope.launch {
                deps.audio.say(line, EngineLines.SUCCESS_RATE, EngineLines.DEFAULT_PITCH)
            }
            return Verdict.ACCEPT
        }
        locked = true
        // [DEVIATION, authorised — iOS D45] Stale prompt; the identical line is
        // in `SinglePickModel.pick`. The set is complete; announcing the task
        // now would cut the success line short and strand the round.
        announceJob?.cancel()
        announceJob = null
        deps.audio.success()
        deps.fireConfetti()
        // Advance only after the LAST success line has played in full (`ok`).
        val next = idxState + 1
        deps.scope.launch {
            val ok = deps.audio.say(line, EngineLines.SUCCESS_RATE, EngineLines.DEFAULT_PITCH)
            if (!ok || !mounted) return@launch
            locked = false
            if (next >= session.size) {
                announceJob?.cancel()
                announceJob = null
                moodState = Mood.CHEER
                earnedState = deps.award(
                    ExerciseId.SOUND_TWINS,
                    level,
                    StarStrip.kept(starList),
                    session.size,
                )
                doneState = true
                deps.scope.launch { deps.audio.say(EngineLines.BRAVO_FOUND) }
            } else {
                moodState = Mood.IDLE
                foundState = emptyList()
                idxState = next
                scheduleAnnounce()
            }
        }
        return Verdict.ACCEPT
    }

    // --- Audio affordances ---------------------------------------------------

    /** The big 🔊 button — locked-guarded, no unlock (the TSX shape). */
    override fun replayPrompt() {
        if (locked || doneState || session.isEmpty()) return
        val line = twinPrompt(session[idxState].family)
        deps.scope.launch { deps.audio.say(line) }
    }

    /**
     * A tile's « Écouter » — speaks the tile's own family's sound
     * (`tile.sound`); locked-guarded in the TSX.
     */
    fun preview(text: String) {
        if (locked) return
        deps.audio.unlock()
        deps.scope.launch { deps.audio.say(text) }
    }

    // --- Announce (350 ms after every round becomes current) -----------------

    private fun scheduleAnnounce() {
        announceJob?.cancel()
        announceJob = null
        if (doneState || session.isEmpty()) return
        val expected = idxState
        val line = twinPrompt(session[expected].family)
        announceJob = deps.scope.launch {
            deps.delay(EngineLines.ANNOUNCE_DELAY_MS)
            if (!isActive || !mounted || doneState || idxState != expected) return@launch
            deps.audio.say(line)
        }
    }
}
